File: LEETCODE/Problem2/AddTwoNumbers.cpp
// Runtime: 28 ms, faster than 69.98% of C++ online submissions for Add Two Numbers.
// Memory Usage: 71.4 MB, less than 47.64% of C++ online submissions for Add Two Numbers.

#include <bits/stdc++.h>

using namespace std;
  
class ListNode 
{
  public:
    int val;
    ListNode *next;

   ListNode(int val)
   {
       this->val=val;
       this->next=NULL;
   }
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *l3=new ListNode(0);
        l3->val=0;
        ListNode *p=l3;
        int carry=0;
        while(l1!=NULL || l2!=NULL)
        {
            int x=0;
            int y=0;
            
            if(l1)
            {
                x=l1->val;
                l1=l1->next;
            }
            if(l2)
            {
                y=l2->val;
                l2=l2->next;
            }

            int val=x+y+carry;

            carry=(int) val/10;
            val=val%10;

            p->next=new ListNode(val);
            p=p->next;
            p->val=val;
            
        }

        if(carry)
        {
            p->next=new ListNode(carry);
            p=p->next;
            p->val=carry;
        }

      return l3->next;
    }
};

int main()
{
    int l1[]={9,9,9,9,9,9,9};
    int l2[]={9,9,9,9};

    ListNode *list1=NULL,*list2=NULL,*p=NULL;

    int n1=sizeof(l1)/sizeof(l1[0]);

    int n2=sizeof(l2)/sizeof(l2[0]);

      
    for(int i=0;i<n1;i++)
    {
        if(i==0)
        {
            list1=new ListNode(l1[0]);
            p=list1;
        }
        else
        {
           p->next=new ListNode(l1[i]);
           p=p->next;
        }
    }

    for(int j=0;j<n2;j++)
    {
        if(j==0)
        {
            list2=new ListNode(l2[0]);
            p=list2;
        }
        else
        {
            p->next=new ListNode(l2[j]);
            p=p->next;
        }
    }
   
    ListNode *l3=NULL;
    Solution sol=Solution();

    l3=sol.addTwoNumbers(list1,list2);

    p=l3;
    while(p!=NULL)
    {
        cout<<p->val<<endl;
        p=p->next;
    }

}

File: LEETCODE/Problem2/AddTwoNumbers.java
// Runtime: 1 ms, faster than 100.00% of Java online submissions for Add Two Numbers.
// Memory Usage: 39.1 MB, less than 86.75% of Java online submissions for Add Two Numbers.


import java.util.*;


class ListNode 
{
        int val;
        ListNode next;
        ListNode()
        {

        }
        ListNode(int val)
        { 
            this.val  = val;
            this.next = null;
        }
}


class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry;
        ListNode l3 = null,p=null;
        l3=new ListNode(0);
        p=l3;

        carry=0;

        while(l1!=null || l2!=null)
        {
           int x=0;
           int y=0;
           if(l1!=null)
           {
               x=l1.val;
               l1=l1.next;
           }
           if(l2!=null)
           {
               y=l2.val;
               l2=l2.next;
           }
           int val=x+y+carry;

           carry=val/10;

           val=val%10;

           p.next=new ListNode(val);


          p=p.next;

        }

       if(carry>0)
       {
          p.next=new ListNode(carry);
       }

        return l3.next;
    }
}


public class AddTwoNumbers
{
     public static void main(String[] args) {
         
        int[] l1=new int[] {9,9,9,9,9,9,9};

        int[] l2=new int[] {9,9,9,9};

        ListNode list1=null,list2=null;
        ListNode p=null;
        
        for(int i=0;i<l1.length;i++)
        {
            if(i==0)
            {
                list1=new ListNode(l1[0]);
                p=list1;
            }
            else
            {
                p.next=new ListNode(l1[i]);
                p=p.next;
            }
        }

        
        for(int i=0;i<l2.length;i++)
        {
            if(i==0)
            {
                list2=new ListNode(l2[0]);
                p=list2;
            }
            else
            {
                p.next=new ListNode(l2[i]);
                p=p.next;
            }
        }

        Solution sol=new Solution();
        ListNode list3=sol.addTwoNumbers(list1, list2);

        p=list3;
        while(p!=null)
        {
            System.out.println(p.val);
            p=p.next;
        }

     }
}